shapely/_flatcoords.py

```diff
@@ -0,0 +1,426 @@
+def _get_arrays_linestring(arr, include_z):
+    # the coords and offsets into the coordinates of the linestrings
+    coords, indices = get_coordinates(arr, return_index=True, include_z=include_z)
+    offsets = _indices_to_offsets(indices, len(arr))
+
+    return coords, (offsets,)
+
+
+def _get_arrays_multilinestring(arr, include_z):
+    # explode/flatten the MultiLineStrings
+    arr_flat, part_indices = get_parts(arr, return_index=True)
+    # the offsets into the multilinestring parts
+    offsets2 = _indices_to_offsets(part_indices, len(arr))
+
+    # the coords and offsets into the coordinates of the linestrings
+    coords, indices = get_coordinates(arr_flat, return_index=True, include_z=include_z)
+    offsets1 = np.insert(np.bincount(indices).cumsum(), 0, 0)
+
+    return coords, (offsets1, offsets2)
+
+
+def _get_arrays_polygon(arr, include_z):
+    # explode/flatten the Polygons into Rings
+    arr_flat, ring_indices = get_rings(arr, return_index=True)
+    # the offsets into the exterior/interior rings of the multipolygon parts
+    offsets2 = _indices_to_offsets(ring_indices, len(arr))
+
+    # the coords and offsets into the coordinates of the rings
+    coords, indices = get_coordinates(arr_flat, return_index=True, include_z=include_z)
+    offsets1 = np.insert(np.bincount(indices).cumsum(), 0, 0)
+
+    return coords, (offsets1, offsets2)
+
+
+def _get_arrays_multipolygon(arr, include_z):
+    # explode/flatten the MultiPolygons
+    arr_flat, part_indices = get_parts(arr, return_index=True)
+    # the offsets into the multipolygon parts
+    offsets3 = _indices_to_offsets(part_indices, len(arr))
+
+    # explode/flatten the Polygons into Rings
+    arr_flat2, ring_indices = get_rings(arr_flat, return_index=True)
+    # the offsets into the exterior/interior rings of the multipolygon parts
+    offsets2 = np.insert(np.bincount(ring_indices).cumsum(), 0, 0)
+
+    # the coords and offsets into the coordinates of the rings
+    coords, indices = get_coordinates(arr_flat2, return_index=True, include_z=include_z)
+    offsets1 = np.insert(np.bincount(indices).cumsum(), 0, 0)
+
+    return coords, (offsets1, offsets2, offsets3)
+
+
+def to_ragged_array(arr, include_z=False):
+    """
+    Converts to ragged array representation using a contiguous array of
+    coordinates and offset arrays.
+
+    This function converts an array of geometries to a ragged array
+    (i.e. irregular array of arrays) of coordinates, represented in memory
+    using a single contiguous array of the coordinates, and in addition
+    0, 1 or up to 3 offset arrays that keep track where each sub-array
+    starts and ends.
+
+    This follows the in-memory layout of the variable size list arrays defined
+    by Apache Arrow, as specified for geometries by the GeoArrow project:
+    https://github.com/geoarrow/geoarrow.
+
+    Parameters
+    ----------
+    arr : array_like
+        Array of geometries (1-dimensional).
+    include_z : bool, default False
```

Comment:
Inserting NaNs may result into trouble when reconstructing the geometry.

I think that this function should expect homogeneous typed geometries. 

Mostly it is better to start out strict with input parsing. You can be nice to the user and fix inhomogeneously typed geometries, but maybe the user rather sees an error in that case so that the user is able to fix problems earlier in its process.

So I'd prefer a check on uniqueness of the `coordinate_dimension` as well (probably excluding the empties).

There are `force_2d` and `force_3d` functions if the user wants to fix inhomogeneous dimensionality

Reply:
Yeah, I would generally be fine with being strict about the dimensionality. 

A problem for implementing that, however, is that GEOS itself is not very strict about it, since they basically always store 3D coords, but then just NaNs for z if 2D. So there are some inconsistencies with empty geometries and checking `has_z`. 

And I don't want to exclude empties, since I want to use those functions to be able to store geometry data to disk, and empties are a fact of life ..

Reply:
We could keep the empties, just without asserting their dimensionality.

I just realized that I am contradicting myself (a few years ago); in the `coordinates.py` module I made the `include_z` exactly like you have here.

Reply:
Sorry, re-reading your comment, you meant "excluding the empties" for when checking the dimensionality (exactly because empties are a bit inconsistent there, as I was unnecessarily explaining ;)). Yes, that could indeed work. 

One question though: currently I default to outputting 2D, even if you have 3D data (which is consistent with `get_coordinates`). Do we keep that (and thus the `include_z` keyword)? Or do we check the first non-empty geometry to "infer" the dimensionality? 
(we probably need to keyword anyway, to be robust against corner cases of all missing / empty arrays where the user wants to have an ensured shape of the result)

Reply:
I think I would prefer inferring the dimensionality except when the user explicitly provides it (so `include_z=None` as default). Indeed to deal with all empty/all missing.

Reply:
So it's quite easy to infer the dimensionality for the default case of `include_z=None` (and raise an error if there is a mixed dimensionality). 

But I would prefer that the user can avoid the performance cost of doing this check by explicitly specifying `include_z=True/False`. But if we still want to be strict about homogeneous dimensionality even when not guessing if it's 2D or 3D, we still need to query the dimensionality anyway to be able to be strict .. (since `get_coordinates` handles a mixture fine, and I think that's also fine for that function)

Reply:
I was typing a reply that such a check would be hardly noticable, but then I did a benchmark, and in fact a dimensionality check would have a significant impact on the `to_ragged_array`:

```
In [5]: x = shapely.linestrings(np.random.random((1000, 10, 2)))
In [6]: %timeit shapely.to_ragged_array(x)
174 µs ± 141 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
In [8]: %timeit np.unique(shapely.get_coordinate_dimension(x))
31.9 µs ± 423 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

The type check has similar impact:

```
In [9]: %timeit np.unique(shapely.get_type_id(x))
31.7 µs ± 833 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
> So I think I would go for:

OK, updated the PR to follow that logic.

For the inference, I went for setting `include_z` to `True` once there is at least one 3D geometry, in which case I indeed don't have to do anything special for the empties

Reply:
Of course, since you mentioned yourself that this is only consistent since GEOS 3.9, this is failing for older GEOS versions.

So two options:

- Skip the failing tests related to include_z inference for older GEOS and leave as is
- Do actually add the `is_empty` check and only infer dimensionality on non-empties (that would have a corner case if you only have singular empty geometries, which then would not infer potentially 3D geometries, but that seems very much a corner case)

Reply:
let’s leave this as is and drop a comment in the docstring for the include_z behaviour.

Reply:
OK, I added a note about this in the `include_z` entry of the docstring. 

I think we can also certainly still fine-tune this in the future.